language/move-vm/runtime/src/native_functions.rs

```diff
@@ -139,6 +139,34 @@ impl<'a, 'b> NativeContext<'a, 'b> {
         self.data_store.events()
     }
 
+    pub fn publish_modules(
```

Comment:
And I don't think there is much benefit (unless I'm missing something) to doing it this way, vs `NativeExtensions`, since you will not be able to access this module during this transaction anyway (as it won't be able to link ahead of time)

Reply:
Might have missed something in your proposal. The data store is internal to the move crate, so there's no way to expose this. The only way to currently publish a module is via the session interface, which isn't available in the context of an executing script(function). Though happy to identify a cleaner path forward?

Reply:
The lazy publishing (like setting a mark in the native extension that some stuff has to been published after return from execution) is definitely the more cumbersome path.  We like to programmatically publish code, and errors resulting from validation should become execution errors. Why not do it like this PR suggests? It is the most natural way.

Reply:
> The only way to currently publish a module is via the session interface, which isn't available in the context of an executing script(function)

But it is available in the session. You can do more than one thing in the session. (more on that below)

> Why not do it like this PR suggests? It is the most natural way.

I worry about threading more and more things into the natives like this. Particularly, I worry about long term implications for performance. For example, we'd like to probably in the future do a lot more in module loading for more efficient execution (we really don't take advantage of type information currently, for example). At the very least, Dario was very big on this direction back in the day. 
I think adding additional module loading in the middle of execution might complicate things, and isn't strictly necessary for this to work. 

I know there are a lot of might/maybe/sorta/somedays here. My point is simply keeping things in the runtime simple will make all sorts of endeavors easier in the future. 

I'm also not entirely convinced this is the "most natural way", or at the very least I would argue its subjective.

> Though happy to identify a cleaner path forward?

To put what I was hinting at more clearly, I would imagine something like 

- Native function adds the module/package bytes to the `NativeContextExtensions`
  - You can do verification in here too if you really want to deserialize and verify the module during runtime (though it is very expensive time wise).
  - Linking will definitely be harder here
- At the end of the transaction, grab the module bytes out of `NativeContextExtensions`
- Call `session.publish_module_bundle` on the bytes

TL;DR we can do this already with `NativeContextExtensions`. I'm not sure that this way presented in this PR is really better or worse in the short term. But in the long term, I worry it might complicate things.